### ISSUE TITLE
`ubi` should very checksums by default

### ISSUE THREAD

**Issue by autarch** [houseabsolute/ubi#84]
If a release has an accompanying checksum file, it'd be nice if `ubi` downloaded the checksums and validated them.

This will require looking at a whole bunch of projects to see what naming scheme they use for these checksums and what the files look like. At least I know how this works for my own [`actions-rust-release`](https://github.com/houseabsolute/actions-rust-release)-based projects.

**Comment by autarch** [comment]
I realized that the SBOM files contain checksums for individual files in the released tarball. Making use of this would be a lot more involved than just checksumming the entire archive file, so I'm going to leave this out for now.

**Comment by autarch** [comment]
I mostly implemented this, but then I wondered if this is really worthwhile. Verifying checksums doesn't really tell us that much. It'd detect the case where a file is corrupted during transfer, but in this case it's likely that either A) it will error when we try to unpack/uncompress it; B) the resulting executable won't run. Also, how often does this sort of corruption happen? I've downloaded many, many, many things over the years and I'm not sure I've ever seen this happen!

A much more useful thing to check would be whether the release matches a cryptographic signature. But AFAICT, most projects don't publish this sort of thing.

Of the projects above, only `goreleaser` does this. They publish a `checksums.txt` file along with a `checksums.pem` file (the cert used to sign), and a `checksums.sig` file (the signature). This is all done using [`cosign`](https://github.com/sigstore/cosign).

In this case, verifying the checksum along with verifying the signing of the checksums file provides a much stronger guarantee that the contents of the release artifact have not been either corrupted or, more importantly, tampered with.

**Comment by autarch** [comment]
https://lib.rs/crates/sigstore

**Comment by ELLIOTTCABLE** [comment]
w.r.t. validating signatures - looks like [`yamlfmt`](https://github.com/google/yamlfmt?tab=readme-ov-file#verifying-release-artifacts) is also publishing signatures with releases.

**Comment by autarch** [comment]
Thanks for the pointer to `yamlfmt`.

There's something weird about the cert they're publishing. I can't get any info from it with openssl:

```
openssl x509 -in ~/Downloads/checksums.txt.pem -text
Could not read certificate from /home/autarch/Downloads/checksums.txt.pem
Unable to load certificate
```

**Comment by scop** [comment]
[aqua-registry](https://github.com/aquaproj/aqua-registry) has info about quite a few projects using cosign. Search for `cosign:` in https://github.com/aquaproj/aqua-registry/blob/main/registry.yaml

In addition to cosign, [SLSA provenance](https://slsa.dev/spec/v1.0/distributing-provenance) is another type of information that can be used to verify artifacts. See `slsa_provenance:` in the aqua registry.yaml, and https://github.com/slsa-framework/slsa-verifier

Yet another similar thing is [GitHub artifact attestations](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds). No entries in aqua registry.yaml yet, but some are starting to appear soonish: https://github.com/aquaproj/aqua-registry/issues/32597

**Comment by Adirelle** [comment]
> There's something weird about the cert they're publishing. I can't get any info from it with openssl:
> 
> ```
> openssl x509 -in ~/Downloads/checksums.txt.pem -text
> Could not read certificate from /home/autarch/Downloads/checksums.txt.pem
> Unable to load certificate
> ```

For some reason, this file is a base64-encoded version of the certificate:  

```
base64 -d checksums.txt.pem | openssl x509 -noout -text
```

**Comment by scop** [comment]
> For some reason, this file is a base64-encoded version of the certificate

This is a known cosign issue, it redundantly base64 encodes pem files by default without indicating having done that in the filename: https://github.com/sigstore/cosign/issues/2666#issuecomment-1406636323

**Comment by risu729** [comment]
On June 6th, GitHub started to generate SHA256 checksums for release assets.
We might be able to use them to verify assets.
https://github.blog/changelog/2025-06-03-releases-now-expose-digests-for-release-assets/

**Comment by autarch** [comment]
I think the auto-generated checksums have the same issue as I noted in an earlier comment, which is that checking them doesn't achieve much. There's nothing to prevent a bad actor from replacing release assets if they compromise a repo. Presumably, GitHub will generate new checksums when that happens.

Yes, as noted, if you record the checksum locally for a given asset, then checking it has some value, but this is a huge pain and almost nobody will do this.

I think schemes which involve some sort of actual signature by the author are more valuable.

**Comment by risu729** [comment]
That's right, I could edit the assets with the same name but with a different digest.
